src/sql.rs

```diff
@@ -0,0 +1,1175 @@
+                params![],
+            )?;
+            sql.execute(
+                "ALTER TABLE chats ADD COLUMN locations_last_sent INTEGER DEFAULT 0;",
+                params![],
+            )?;
+            sql.execute(
+                "CREATE INDEX chats_index3 ON chats (locations_send_until);",
+                params![],
+            )?;
+            dbversion = 53;
+            sql.set_config_int(context, "dbversion", 53);
+        }
+        if dbversion < 54 {
+            info!(context, 0, "[migration] v54");
+            sql.execute(
+                "ALTER TABLE msgs ADD COLUMN location_id INTEGER DEFAULT 0;",
+                params![],
+            )?;
+            sql.execute("CREATE INDEX msgs_index6 ON msgs (location_id);", params![])?;
+            dbversion = 54;
+            sql.set_config_int(context, "dbversion", 54);
+        }
+        if dbversion < 55 {
+            sql.execute(
+                "ALTER TABLE locations ADD COLUMN independent INTEGER DEFAULT 0;",
+                params![],
+            )?;
+
+            sql.set_config_int(context, "dbversion", 55);
+        }
+
+        if 0 != recalc_fingerprints {
+            sql.query_map(
+                "SELECT addr FROM acpeerstates;",
+                params![],
+                |row| row.get::<_, String>(0),
+                |addrs| {
+                    for addr in addrs {
+                        if let Some(ref mut peerstate) = Peerstate::from_addr(context, sql, &addr?)
+                        {
+                            peerstate.recalc_fingerprint();
+                            peerstate.save_to_db(sql, false);
+                        }
+                    }
+                    Ok(())
+                },
+            )?;
+        }
+        if 0 != update_file_paths {
+            // versions before 2018-08 save the absolute paths in the database files at "param.f=";
+            // for newer versions, we copy files always to the blob directory and store relative paths.
+            // this snippet converts older databases and can be removed after some time.
+
+            info!(context, 0, "[open] update file paths");
+
+            let repl_from = sql
+                .get_config(context, "backup_for", Some(as_str(context.get_blobdir())))
+                .unwrap();
+
+            let repl_from = dc_ensure_no_slash_safe(&repl_from);
+            sql.execute(
+                &format!(
+                    "UPDATE msgs SET param=replace(param, 'f={}/', 'f=$BLOBDIR/')",
+                    repl_from
+                ),
+                NO_PARAMS,
+            )?;
+
+            sql.execute(
+                &format!(
+                    "UPDATE chats SET param=replace(param, 'i={}/', 'i=$BLOBDIR/');",
+                    repl_from
+                ),
+                NO_PARAMS,
+            )?;
+
+            sql.set_config(context, "backup_for", None);
+        }
+    }
```

Comment:
Nice that you managed to get rid of the goto/match-random-int bits!  I didn't check if the remaining logic is the same, let me know if you'd like me to look at this more carefully.